src/client/dfuse/ops/readdir.c

```diff
@@ -24,240 +24,137 @@
 #include "dfuse_common.h"
 #include "dfuse.h"
 
-/* TODO: This implementation is not complete, in particular it does not
- * correctly handle calls where offset != 0, which potentially generates
- * incorrect results if the number of files is greater than can be observed in
- * one call (approx 25).
- *
- * If the filesystem is only modified from one client then the results should
- * however be correct.
- */
-
 #define LOOP_COUNT 10
 
+struct rd_data {
+	fuse_req_t	req;
+	struct dfuse_inode_entry *inode;
+	daos_anchor_t	*anchor;
+	void		*buf;
+	size_t		size;
+	size_t		b_offset;
+};
+
+int
+rd_cb(dfs_t *dfs, dfs_obj_t *dir, const char name[], void *_udata)
+{
+	struct rd_data	*udata = (struct rd_data *)_udata;
+	struct dfuse_projection_info *fs_handle = fuse_req_userdata(udata->req);
+	dfs_obj_t	*obj;
+	daos_obj_id_t	oid;
+	struct stat	stbuf = {0};
+	int		ns;
+	int		rc;
+
+	/*
+	 * MSC - from fuse fuse_add_direntry: "From the 'stbuf' argument the
+	 * st_ino field and bits 12-15 of the st_mode field are used. The other
+	 * fields are ignored." So we only need to lookup the entry for the
+	 * mode.
+	 */
```

Comment:
Arrr, this means an extra RPC for each entry in the directory? That's pretty bad.
We should consider extending the API to fetch extra args as part of listing.

Reply:
yes..
well we would need to extend the actual DAOS API to list + retrieve the value of each associated key. that means a generic API for dkey listing to retrieve the akeys + single/array values under each akey vs what we do now with list then fetch.